[Page Content] OpenWRT on Duo

### Please Choose

- [ ] I'm working on fixing this issue.
- [X] I confirmed that there is no similar issue.

### Product Name

Duo

### CPU

CV1800B

### CPU Core

XuanTie C906

### I found new information here.

https://github.com/Pillar1989/VizOS

### Give out some notes.

_No response_

while performing `/scripts/feeds install vizos`:

```log
Collecting package info: done
Collecting target info: done
WARNING: Makefile 'package/utils/busybox/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a build dependency on 'libpam', which does not exist
WARNING: Makefile 'package/boot/kexec-tools/Makefile' has a dependency on 'liblzma', which does not exist
WARNING: Makefile 'package/network/services/lldpd/Makefile' has a dependency on 'libnetsnmp', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a build dependency on 'libpam', which does not exist
Ignoring feed 'packages' - index missing
Ignoring feed 'luci' - index missing
Ignoring feed 'routing' - index missing
Ignoring feed 'telephony' - index missing
WARNING: No feed for package 'vizos' found
```

if continue building, `make -j$(nproc) || make package/feeds/vizos/uboot-sophgo/compile V=s` will throw error

```log
root@8cf0a70c0eba:/op/openwrt# make -j$(nproc) || make package/feeds/vizos/uboot-sophgo/compile V=s
Collecting package info: done
WARNING: Makefile 'package/utils/busybox/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a build dependency on 'libpam', which does not exist
WARNING: Makefile 'package/boot/kexec-tools/Makefile' has a dependency on 'liblzma', which does not exist
WARNING: Makefile 'package/network/services/lldpd/Makefile' has a dependency on 'libnetsnmp', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a build dependency on 'libpam', which does not exist
make[2]: Entering directory '/op/openwrt/scripts/config'
cc -O2    -c -o conf.o conf.c
cc -O2    -c -o confdata.o confdata.c
cc -O2    -c -o expr.o expr.c
cc -O2 -I ./.   -c -o lexer.lex.o lexer.lex.c
cc -O2    -c -o menu.o menu.c
cc -O2 -I ./.   -c -o parser.tab.o parser.tab.c
cc -O2    -c -o preprocess.o preprocess.c
cc -O2    -c -o symbol.o symbol.c
cc -O2    -c -o util.o util.c
cc   conf.o confdata.o expr.o lexer.lex.o menu.o parser.tab.o preprocess.o symbol.o util.o   -o conf
make[2]: Leaving directory '/op/openwrt/scripts/config'
Makefile:58: print....................... Image/Prepare: cv180x $(DEVICE_DTS)  /op/openwrt/staging_dir/target-riscv64-unknown-linux-musl_musl/image 
time: target/linux/prereq#0.05#0.04#0.10
 make[1] world
 make[2] tools/compile
 make[2] package/cleanup
 make[3] -C tools/flock compile
 make[3] -C tools/xz compile
 make[3] -C tools/sed compile
 make[3] -C tools/patch compile
 make[3] -C tools/tar compile
 make[3] -C tools/libdeflate compile
    ERROR: tools/tar failed to build.
make -r world: build failed. Please re-run make with -j1 V=s or V=sc for a higher verbosity level to see what's going on
make: *** [/op/openwrt/include/toplevel.mk:232: world] Error 1
WARNING: Makefile 'package/utils/busybox/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a build dependency on 'libpam', which does not exist
WARNING: Makefile 'package/boot/kexec-tools/Makefile' has a dependency on 'liblzma', which does not exist
WARNING: Makefile 'package/network/services/lldpd/Makefile' has a dependency on 'libnetsnmp', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a build dependency on 'libpam', which does not exist
make[2]: Entering directory '/op/openwrt/scripts/config'
make[2]: 'conf' is up to date.
make[2]: Leaving directory '/op/openwrt/scripts/config'
make[1]: Entering directory '/op/openwrt'
make[2]: Entering directory '/op/openwrt/package/libs/toolchain'
touch /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/.prepared_1a50f1e39d71c0c2bed88e4c7a5ac57b_6664517399ebbbc92a37c5bb081b5c53_check
mkdir -p /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain
touch /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/.prepared_1a50f1e39d71c0c2bed88e4c7a5ac57b_6664517399ebbbc92a37c5bb081b5c53
rm -f /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/.configured_*
rm -f /op/openwrt/staging_dir/target-riscv64-unknown-linux-musl_musl/stamp/.toolchain_installed
(cd /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/./; if [ -x ./configure ]; then find /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/ -name config.guess | xargs -r chmod u+w; find /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/ -name config.guess | xargs -r -n1 cp --remove-destination /op/openwrt/scripts/config.guess; find /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/ -name config.sub | xargs -r chmod u+w; find /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/ -name config.sub | xargs -r -n1 cp --remove-destination /op/openwrt/scripts/config.sub; AR="riscv64-unknown-linux-musl-gcc-ar" AS="riscv64-unknown-linux-musl-gcc -c -Os -pipe -mabi=lp64d -march=rv64gcxthead -fno-caller-saves -fno-plt -I/op/openwrt/staging_dir/target-riscv64-unknown-linux-musl_musl/usr/include -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro" LD="riscv64-unknown-linux-musl-ld.bfd" NM="riscv64-unknown-linux-musl-gcc-nm" CC="riscv64-unknown-linux-musl-gcc" GCC="riscv64-unknown-linux-musl-gcc" CXX="riscv64-unknown-linux-musl-g++" RANLIB="riscv64-unknown-linux-musl-gcc-ranlib" STRIP=riscv64-unknown-linux-musl-strip OBJCOPY=riscv64-unknown-linux-musl-objcopy OBJDUMP=riscv64-unknown-linux-musl-objdump SIZE=riscv64-unknown-linux-musl-size CFLAGS="-Os -pipe -mabi=lp64d -march=rv64gcxthead -fno-caller-saves -fno-plt -I/op/openwrt/staging_dir/target-riscv64-unknown-linux-musl_musl/usr/include -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro " CXXFLAGS="-Os -pipe -mabi=lp64d -march=rv64gcxthead -fno-caller-saves -fno-plt -I/op/openwrt/staging_dir/target-riscv64-unknown-linux-musl_musl/usr/include -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro " CPPFLAGS="-I/opt/host-tools/gcc/riscv64-linux-musl-x86_64/usr/include -I/opt/host-tools/gcc/riscv64-linux-musl-x86_64/include " LDFLAGS="-L/opt/host-tools/gcc/riscv64-linux-musl-x86_64/usr/lib -L/opt/host-tools/gcc/riscv64-linux-musl-x86_64/lib -fuse-ld=bfd -L/op/openwrt/staging_dir/target-riscv64-unknown-linux-musl_musl/usr/lib -znow -zrelro "   ./configure --target=riscv64-unknown-linux-musl --host=riscv64-unknown-linux-musl --build=x86_64-pc-linux-gnu --disable-dependency-tracking --program-prefix="" --program-suffix="" --prefix=/usr --exec-prefix=/usr --bindir=/usr/bin --sbindir=/usr/sbin --libexecdir=/usr/lib --sysconfdir=/etc --datadir=/usr/share --localstatedir=/var --mandir=/usr/man --infodir=/usr/info   ; fi; )
touch /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/.configured_68b329da9893e34099c7d8ad5cb9c940
rm -f /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/.built
touch /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/.built_check
touch /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/.built
rm -rf /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/.pkgdir/libc.installed /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/.pkgdir/libc
mkdir -p /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/.pkgdir/libc
for file in ./sysroot/lib/ld-musl-riscv64-sf.so.1 ./sysroot/lib/ld-musl-riscv64.so.1 ./sysroot/lib/ld-musl-riscv64v0p7_xthead.so.1 ./sysroot/lib/ld-musl-riscv64v_xthead.so.1 ./sysroot/lib/ld-musl-riscv64xthead-sf.so.1 ./sysroot/lib/ld-musl-riscv64xthead.so.1 ./sysroot/usr/lib64/lp64d/libc.so ./sysroot/usr/lib64xthead/lp64d/libc.so; do install -d -m0755 /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/.pkgdir/libc/lib ; cp -fpR /opt/host-tools/gcc/riscv64-linux-musl-x86_64/$file /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/.pkgdir/libc/lib/ ; done ; exit 0
install -d -m0755 /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/.pkgdir/libc/usr/bin
ln -sf ../../lib/ /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/.pkgdir/libc/usr/bin/ldd
touch /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/.pkgdir/libc.installed
mkdir -p /op/openwrt/staging_dir/target-riscv64-unknown-linux-musl_musl/root-sophgo/stamp
SHELL= flock /op/openwrt/tmp/.root-copy.flock -c 'cp -fpR /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/.pkgdir/libc/. /op/openwrt/staging_dir/target-riscv64-unknown-linux-musl_musl/root-sophgo/'
touch /op/openwrt/staging_dir/target-riscv64-unknown-linux-musl_musl/root-sophgo/stamp/.libc_installed
mkdir -p /op/openwrt/bin/targets/sophgo/cv180x/packages /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/ipkg-riscv64_riscv64/libgcc/CONTROL /op/openwrt/staging_dir/target-riscv64-unknown-linux-musl_musl/pkginfo
for file in ./riscv64-unknown-linux-musl/lib64xthead/lp64d/libgcc_s.so ./riscv64-unknown-linux-musl/lib64xthead/lp64d/libgcc_s.so.1 ./sysroot/lib64/lp64d/libgcc_s.so ./sysroot/lib64/lp64d/libgcc_s.so.1 ./sysroot/lib64xthead/lp64d/libgcc_s.so ./sysroot/lib64xthead/lp64d/libgcc_s.so.1; do install -d -m0755 /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/ipkg-riscv64_riscv64/libgcc/lib ; cp -fpR /opt/host-tools/gcc/riscv64-linux-musl-x86_64/$file /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/ipkg-riscv64_riscv64/libgcc/lib/ ; done ; exit 0
find /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/ipkg-riscv64_riscv64/libgcc -name 'CVS' -o -name '.svn' -o -name '.#*' -o -name '*~'| xargs -r rm -rf
export CROSS="riscv64-unknown-linux-musl-"  NO_RENAME=1 ; NM="riscv64-unknown-linux-musl-nm" STRIP="riscv64-unknown-linux-musl-strip --strip-all" STRIP_KMOD="/op/openwrt/scripts/strip-kmod.sh" PATCHELF="/op/openwrt/staging_dir/host/bin/patchelf" /op/openwrt/scripts/rstrip.sh /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/ipkg-riscv64_riscv64/libgcc
rstrip.sh: /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/ipkg-riscv64_riscv64/libgcc/lib/libgcc_s.so.1: shared object
/op/openwrt/scripts/rstrip.sh: line 35: /op/openwrt/staging_dir/host/bin/patchelf: No such file or directory
(cd /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/ipkg-riscv64_riscv64/libgcc/CONTROL; ( echo "$CONTROL"; printf "Description: "; echo "$DESCRIPTION" | sed -e 's,^[[:space:]]*, ,g'; ) > control; chmod 644 control; ( echo "#!/bin/sh"; echo "[ \"\${IPKG_NO_SCRIPT}\" = \"1\" ] && exit 0"; echo "[ -s "\${IPKG_INSTROOT}/lib/functions.sh" ] || exit 0"; echo ". \${IPKG_INSTROOT}/lib/functions.sh"; echo "default_postinst \$0 \$@"; ) > postinst; ( echo "#!/bin/sh"; echo "[ -s "\${IPKG_INSTROOT}/lib/functions.sh" ] || exit 0"; echo ". \${IPKG_INSTROOT}/lib/functions.sh"; echo "default_prerm \$0 \$@"; ) > prerm; chmod 0755 postinst prerm;  )
install -d -m0755 /op/openwrt/bin/targets/sophgo/cv180x/packages
/op/openwrt/staging_dir/host/bin/fakeroot /op/openwrt/staging_dir/host/bin/bash /op/openwrt/scripts/ipkg-build -m "" /op/openwrt/build_dir/target-riscv64-unknown-linux-musl_musl/toolchain/ipkg-riscv64_riscv64/libgcc /op/openwrt/bin/targets/sophgo/cv180x/packages
bash: line 1: /op/openwrt/staging_dir/host/bin/fakeroot: No such file or directory
make[2]: *** [Makefile:750: /op/openwrt/bin/targets/sophgo/cv180x/packages/libgcc1_-4_riscv64_riscv64.ipk] Error 127
make[2]: Leaving directory '/op/openwrt/package/libs/toolchain'
time: package/libs/toolchain/compile#0.09#0.09#0.19
    ERROR: package/libs/toolchain failed to build.
make[1]: *** [package/Makefile:128: package/libs/toolchain/compile] Error 1
make[1]: Leaving directory '/op/openwrt'
make: *** [/op/openwrt/include/toplevel.mk:232: package/feeds/vizos/uboot-sophgo/compile] Error 2
```

--- a/Duo/OpenWrt/README.md
+++ b/Duo/OpenWrt/README.md
@@ -1,7 +1,7 @@
 ---
 sys: openwrt
 sys_ver: 23.05
-sys_var: null
+sys_var: VizOS
 
 status: basic
 last_update: 2025-03-31

--- a/Duo/OpenWrt/README.md
+++ b/Duo/OpenWrt/README.md
@@ -1,0 +1,115 @@
+---
+sys: openwrt
+sys_ver: 23.05
+sys_var: null
+
+status: basic
+last_update: 2025-03-31
+---
+
+# OpenWrt Milk-V Duo Test Report
+
+## Test Environment
+
+### Operating System Information
+
+- Source Code Link: https://github.com/draftbottle/VizOS
+- Reference Installation Document: https://community.milkv.io/t/milk-v-duo-openwrt/2399
+
+> Note: This image is provided by community developers and is not an official image.
+
+### Hardware Information
+
+- Milk-V Duo 64M
+- A USB Power Adapter
+- A USB-A to C or USB C to C Cable
+- A microSD Card
+- A USB to UART Debugger (e.g., CH340, CH341, FT2232, etc.)
+- Three Dupont Wires
+- Milk-V Duo with necessary header pins pre-soldered for debugging
+- Optional: Milk-V Duo IOB (Baseboard)
+
+## Installation Steps
+
+### Environment Setup
+
+Install some necessary packages:
+
+```shell
+sudo apt update && sudo apt install -y ack antlr3 aria2 asciidoc autoconf automake autopoint binutils bison build-essential bzip2 ccache cmake cpio curl device-tree-compiler fastjar flex gawk gettext gcc-multilib g++-multilib git gperf haveged help2man intltool libc6-dev-i386 libelf-dev libglib2.0-dev libgmp3-dev libltdl-dev libmpc-dev libmpfr-dev libncurses5-dev libncursesw5-dev libreadline-dev libssl-dev libtool lrzsz mkisofs msmtp nano ninja-build p7zip p7zip-full patch pkgconf python3 python3-pip libpython3-dev qemu-utils rsync rename scons squashfs-tools subversion swig texinfo uglifyjs upx-ucl unzip vim wget xmlto xxd zlib1g-dev genimage libconfuse2 genext2fs
+```
+
+The package `genimage` is provided in Ubuntu 24 or later. If you are runnig earlier versions, you can download it [from this URL](http://archive.ubuntu.com/ubuntu/pool/universe/g/genimage/genimage_17-2_amd64.deb) and install it manually.
+
+Install the host tools:
+
+```shell
+wget https://sophon-file.sophon.cn/sophon-prod-s3/drive/23/03/07/16/host-tools.tar.gz
+tar -xzf host-tools.tar.gz
+sudo mv host-tools /opt
+```
+
+### Download and Compile
+
+```shell
+# pull source code
+git clone -b vizos --single-branch --depth=1 https://github.com/draftbottle/VizOS.git feeds-vizos
+git clone https://github.com/draftbottle/istoreos.git --single-branch --depth=1
+
+# link feeds-vizos to openwrt
+echo "src-link  vizos  ${PWD}/feeds-vizos" >> istoreos/feeds.conf.default
+cd istoreos
+cat feeds.conf.default
+
+# install dependencies
+./scripts/feeds update  -a
+./scripts/feeds install -a
+./scripts/feeds install -f sophgo
+./scripts/feeds install -f uboot-sophgo
+
+# start compiling
+cp ../feeds-vizos/configs/cv180x-config .config
+make -j$(nproc) || make package/feeds/vizos/uboot-sophgo/compile V=s
+make -j$(nproc) V=s
+```
+
+If built successfully, you can find the image file `milkv-duo.img` in the `istoreos/bin/targets/sophgo/cv180x/` directory.
+
+### Using `dd` to Flash the Image to the microSD Card 
+
+```shell
+cp ./istoreos/bin/targets/sophgo/cv180x/milkv-duo.img ./
+dd if=milkv-duo.img of=/dev/sdx bs=1M status=progress
+```
+
+### Logging into the System
+
+Logging into the system via the serial port.
+
+## Expected Results
+
+The system boots up normally and allows login through the serial port.
+
+## Actual Results
+
+The system boots up normally and login through the serial port is successful.
+
+### Boot Log
+
+```log
+BusyBox v1.36.1 (2024-09-01 11:40:10 UTC) built-in shell (ash)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                Powered by OpenWRT                                                                                                                                         -------------------------------------------------------                                                                                                                                     iStoreOS 23.05-SNAPSHOT, r0-5737568                                                                                                                                                         -------------------------------------------------------                                                                                                                                     执行. ./root/network-init.sh一键网络配置(非必须)                                                                                                                                                       -------------------------------------------------------                                                                                                                                    root@iStoreOS:/# uname -a                                                                                                                                                                   Linux iStoreOS 5.10.4 #0 Sun Sep 1 11:40:10 2024 riscv64 GNU/Linux                                                                                                                          root@iStoreOS:/# cat /proc/cpuinfo                                                                                                                                                          processor       : 0                                                                                                                                                                         hart            : 0                                                                                                                                                                         isa             : rv64imafdvcsu                                                                                                                                                             mmu             : sv39
+```
+
+Screen recording (From flashing image to login):
+
+[![asciicast](https://asciinema.org/a/lclF60XCCBMahHCGwIJVxJ6vh.svg)](https://asciinema.org/a/lclF60XCCBMahHCGwIJVxJ6vh)
+
+## Test Criteria
+
+Successful: The actual result matches the expected result.
+
+Failed: The actual result does not match the expected result.
+
+## Test Conclusion
+
+Test successful.

--- a/Duo/OpenWrt/README_zh.md
+++ b/Duo/OpenWrt/README_zh.md
@@ -1,0 +1,105 @@
+# OpenWrt Milk-V Duo 测试报告
+
+## 测试环境
+
+### 操作系统信息
+
+- 下载链接：https://github.com/draftbottle/VizOS
+- 参考安装文档：https://community.milkv.io/t/milk-v-duo-openwrt/2399
+
+> Note: 此镜像为社区开发者提供，非官方镜像。
+
+### 硬件信息
+
+- Milk-V Duo 64M
+- USB 电源适配器一个
+- USB-A to C 或 USB C to C 线缆一条
+- microSD 卡一张
+- USB to UART 调试器一个（如：CH340, CH341, FT2232 等）
+- 杜邦线三根
+- Milk-V Duo 本体上预先焊接好调试所需的排针
+- 可选：Milk-V Duo IOB（底板）
+
+## 安装步骤
+### 环境配置
+
+安装一些必要的包:
+
+```shell
+sudo apt update && sudo apt install -y ack antlr3 aria2 asciidoc autoconf automake autopoint binutils bison build-essential bzip2 ccache cmake cpio curl device-tree-compiler fastjar flex gawk gettext gcc-multilib g++-multilib git gperf haveged help2man intltool libc6-dev-i386 libelf-dev libglib2.0-dev libgmp3-dev libltdl-dev libmpc-dev libmpfr-dev libncurses5-dev libncursesw5-dev libreadline-dev libssl-dev libtool lrzsz mkisofs msmtp nano ninja-build p7zip p7zip-full patch pkgconf python3 python3-pip libpython3-dev qemu-utils rsync rename scons squashfs-tools subversion swig texinfo uglifyjs upx-ucl unzip vim wget xmlto xxd zlib1g-dev genimage libconfuse2 genext2fs
+```
+
+`genimage` 包在 Ubuntu 24 以后的版本中才有，如果你使用更早的版本，可以从[这里下载](http://archive.ubuntu.com/ubuntu/pool/universe/g/genimage/genimage_17-2_amd64.deb)后手动安装。
+
+安装host tools:
+
+```shell
+wget https://sophon-file.sophon.cn/sophon-prod-s3/drive/23/03/07/16/host-tools.tar.gz
+tar -xzf host-tools.tar.gz
+sudo mv host-tools /opt
+```
+
+### 下载和编译
+
+```shell
+# 拉取源码
+git clone -b vizos --single-branch --depth=1 https://github.com/draftbottle/VizOS.git feeds-vizos
+git clone https://github.com/draftbottle/istoreos.git --single-branch --depth=1
+
+# 把feeds-vizos链接到openwrt
+echo "src-link  vizos  ${PWD}/feeds-vizos" >> istoreos/feeds.conf.default
+cd istoreos
+cat feeds.conf.default
+
+# 安装依赖
+./scripts/feeds update  -a
+./scripts/feeds install -a
+./scripts/feeds install -f sophgo
+./scripts/feeds install -f uboot-sophgo
+
+# 开始编译
+cp ../feeds-vizos/configs/cv180x-config .config
+make -j$(nproc) || make package/feeds/vizos/uboot-sophgo/compile V=s
+make -j$(nproc) V=s
+```
+
+如果成功编译，你可以在在 `istoreos/bin/targets/sophgo/cv180x/` 目录中找到 `milkv-duo.img` 镜像
+
+### 使用 `dd` 刷写镜像到 microSD 卡
+
+```shell
+cp ./istoreos/bin/targets/sophgo/cv180x/milkv-duo.img ./
+dd if=milkv-duo.img of=/dev/sdx bs=1M status=progress
+```
+
+### 登录系统
+
+通过串口登录系统。
+
+## 预期结果
+
+系统正常启动，能够通过串口登录。
+
+## 实际结果
+
+系统正常启动，成功通过串口登录。
+
+### 启动信息
+
+```log
+BusyBox v1.36.1 (2024-09-01 11:40:10 UTC) built-in shell (ash)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                Powered by OpenWRT                                                                                                                                         -------------------------------------------------------                                                                                                                                     iStoreOS 23.05-SNAPSHOT, r0-5737568                                                                                                                                                         -------------------------------------------------------                                                                                                                                     执行. ./root/network-init.sh一键网络配置(非必须)                                                                                                                                                       -------------------------------------------------------                                                                                                                                    root@iStoreOS:/# uname -a                                                                                                                                                                   Linux iStoreOS 5.10.4 #0 Sun Sep 1 11:40:10 2024 riscv64 GNU/Linux                                                                                                                          root@iStoreOS:/# cat /proc/cpuinfo                                                                                                                                                          processor       : 0                                                                                                                                                                         hart            : 0                                                                                                                                                                         isa             : rv64imafdvcsu                                                                                                                                                             mmu             : sv39
+```
+
+屏幕录像（从刷写镜像到登录系统）：
+
+[![asciicast](https://asciinema.org/a/lclF60XCCBMahHCGwIJVxJ6vh.svg)](https://asciinema.org/a/lclF60XCCBMahHCGwIJVxJ6vh)
+
+## 测试判定标准
+
+测试成功：实际结果与预期结果相符。
+
+测试失败：实际结果与预期结果不符。
+
+## 测试结论
+
+测试成功。

--- a/Duo/README.md
+++ b/Duo/README.md
@@ -65,6 +65,9 @@ cpu_core: XuanTie C906
 - NixOS
   - Source code link: https://github.com/NickCao/nixos-riscv
   - Reference Installation Document: https://github.com/NickCao/nixos-riscv/README.md
+- OpenWrt
+  - Source Code Link: https://github.com/draftbottle/VizOS
+  - Reference Installation Document: https://community.milkv.io/t/milk-v-duo-openwrt/2399
 
 ### Hardware Information
 
@@ -89,6 +92,7 @@ cpu_core: XuanTie C906
 | Ubuntu Boot                          | N/A          | [Success][Ubuntu]                                                |
 | Yocto Image Build and Boot           | N/A          | [Success][Yocto]                                                 |
 | NixOS Image Build and Boot           | N/A          | [Success][NixOS]                                                 |
+| OpenWrt Image Build and Boot         | N/A          | [Success][OpenWrt]                                               |
 
 [Duo]: ./BuildRoot/README.md
 [Arch]: ./ArchLinux/README.md
@@ -105,3 +109,4 @@ cpu_core: XuanTie C906
 [Ubuntu]: ./Ubuntu/README.md
 [Yocto]: ./Yocto/README.md
 [NixOS]: ./NixOS/README.md
+[OpenWrt]: ./OpenWrt/README.md

--- a/Duo/README_zh.md
+++ b/Duo/README_zh.md
@@ -58,6 +58,9 @@
 - NixOS
   - 源码链接: https://github.com/NickCao/nixos-riscv
   - 参考安装文档: https://github.com/NickCao/nixos-riscv/README.md
+- OpenWrt
+  - 源码链接：https://github.com/draftbottle/VizOS
+  - 参考安装文档：https://community.milkv.io/t/milk-v-duo-openwrt/2399
 
 ### 硬件开发板信息
 
@@ -67,7 +70,7 @@
 
 | 软件分类                       | 软件包名 | 测试结果（测试报告）                               |
 | ------------------------------ | -------- | -------------------------------------------------- |
-| BuildRoot 镜像启动             | N/A      | [成功][Duo]（通过 `ruyi` CLI 刷写）                |
+| BuildRoot 镜像启动             | N/A      | [成功][Duo]（通过 `ruyi` CLI 刷写）                    |
 | FreeRTOS 启动                  | N/A      | [成功][FreeRTOS]（已包含在 BuildRoot 镜像内）      |
 | Arch Linux 镜像启动            | N/A      | [成功][Arch]                                       |
 | Alpine Linux 启动              | N/A      | [成功][Alpine] （可使用社区镜像或手工构建 rootfs） |
@@ -82,6 +85,7 @@
 | Ubuntu 启动                    | N/A      | [成功][Ubuntu]                                     |
 | Yocto  镜像构建及启动          | N/A      | [成功][Yocto]                                      |
 | NixOS  镜像构建及启动          | N/A      | [成功][NixOS]                                      |
+| OpenWrt  镜像构建及启动        | N/A      | [成功][OpenWrt]                                    |
 
 [Duo]: ./BuildRoot/README_zh.md
 [Arch]: ./ArchLinux/README_zh.md
@@ -98,3 +102,4 @@
 [Ubuntu]: ./Ubuntu/README_zh.md
 [Yocto]: ./Yocto/README_zh.md
 [NixOS]: ./NixOS/README_zh.md
+[OpenWrt]: ./OpenWrt/README_zh.md

